RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+            if (r != 1)
+              printf(", ");
+            if (r <= 0) {
+              printf("Invalid rhNum (%d) for get RhFracIdx!\n", r);
+            }
+            printf("%d", rechits[r - 1]);
+          }
+          printf(")\n\n");
+        }
+      }
+      alpaka::syncBlockThreads(acc);  // all (or none) threads call sync
+    }
+
+    // Set initial cluster position (energy) to seed rechit position (energy)
+    if (tid < nSeeds) {
+      int i = getSeedRhIdx(seeds, tid);
+      clusterPos[tid] = Position4{pfRecHits[i].x(), pfRecHits[i].y(), pfRecHits[i].z(), 1.0};
+      prevClusterPos[tid] = clusterPos[tid];
+      clusterEnergy[tid] = pfRecHits[i].energy();
+      for (int r = 0; r < (nRHNotSeed - 1); r++) {
+        fracView[pfClusteringVars[i].seedFracOffsets() + r + 1].pfrhIdx() = rechits[r];
+        fracView[pfClusteringVars[i].seedFracOffsets() + r + 1].frac() = -1.;
+      }
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    int rhThreadIdx = -1;
+    Position4 rhThreadPos;
+    if (tid < (nRHNotSeed - 1)) {
+      rhThreadIdx = rechits[tid];  // Index when thread represents rechit
+      rhThreadPos = Position4{pfRecHits[rhThreadIdx].x(), pfRecHits[rhThreadIdx].y(), pfRecHits[rhThreadIdx].z(), 1.};
+    }
+
+    // Neighbors when threadIdx represents seed
+    int seedThreadIdx = -1;
+    Neighbours4 seedNeighbors = Neighbours4{-9, -9, -9, -9};
+    float seedEnergy = -1.;
+    Position4 seedInitClusterPos = Position4{0., 0., 0., 0.};
+    if (tid < nSeeds) {
+      if constexpr (debug)
+        printf("tid: %d\n", tid);
+      seedThreadIdx = getSeedRhIdx(seeds, tid);
+      seedNeighbors = Neighbours4{pfRecHits[seedThreadIdx].neighbours()(0),
+                                  pfRecHits[seedThreadIdx].neighbours()(1),
+                                  pfRecHits[seedThreadIdx].neighbours()(2),
+                                  pfRecHits[seedThreadIdx].neighbours()(3)};
+      seedEnergy = pfRecHits[seedThreadIdx].energy();
+
+      // Compute initial cluster position shift for seed
+      computeClusterPos(pfClusParams, seedInitClusterPos, 1., seedThreadIdx, pfRecHits, rhENormInv);
+    }
+
+    do {
+      if constexpr (debug) {
+        if (once_per_block(acc))
+          printf("\n--- Now on iter %d for topoId %d ---\n", iter, topoId);
+      }
+
+      // Reset rhFracSum
+      rhFracSum[tid] = 0.;
+      if (once_per_block(acc))
+        diff2 = -1;
+
+      if (tid < (nRHNotSeed - 1)) {
+        for (int s = 0; s < nSeeds; s++) {
+          float dist2 = (clusterPos[s].x - rhThreadPos.x) * (clusterPos[s].x - rhThreadPos.x) +
+                        (clusterPos[s].y - rhThreadPos.y) * (clusterPos[s].y - rhThreadPos.y) +
+                        (clusterPos[s].z - rhThreadPos.z) * (clusterPos[s].z - rhThreadPos.z);
+
+          float d2 = dist2 / pfClusParams.showerSigma2();
+          float fraction = clusterEnergy[s] * rhENormInv * expf(-0.5 * d2);
```

Comment:
Is there any way to store `d2` and `fraction` instead of recomputing them below ?
Or, is recomputing them faster than storing them ?

Reply:
I think it is a memory issue. If I try to use shared memory to store these (2 `float` arrays that are `nSeeds * threadsPerBlockForClustering` long), I run into a compiler error regarding the shared memory limit. Potentially they can be stored as members of a temporary SoA? Maybe those SoA accesses are sufficiently fast. Is this something you'd like me to pursue further?

Reply:
I don't know if using a temporary SoA will be faster or slower - we can look into it later if it turns out that we need to optimise this kernel.